src/operation.rs

```diff
@@ -106,11 +128,12 @@ pub(crate) trait Operation {
     fn extract_at_cluster_time(&self, _response: &RawDocument) -> Result<Option<Timestamp>>;
 
     /// Interprets the server response to the command.
-    fn handle_response(
-        &self,
+    fn handle_response<'a>(
+        &'a self,
         response: RawCommandResponse,
-        description: &StreamDescription,
-    ) -> Result<Self::O>;
+        description: &'a StreamDescription,
+        session: Option<&'a mut ClientSession>,
+    ) -> OperationResponse<'a, Self::O>;
```

Comment:
Yup, that SGTM.  Thanks for linking those cases - I think that actually supports doing this since (as you said) none of those were actual blockers, just it not being clear where the bounds came from.  If we have the bounds upfront on `Collection` that'll remove this sort of confusion.

Reply:
Idea (and sorry this came to me so late):
* `Operation` defines `handle_response_async` (as you've got it in this PR, just with a rename)
* `OperationWithDefaults` defines both `handle_response` (with a no-op impl) and `handle_response_async` (with an impl that just wraps the sync `handle_response` in `BoxFuture`)
* `impl Operation for OperationWithDefaults` uses `OperationWithDefaults::handle_response_async` as the body of `Operation::handle_response_async`
* Operation execution only calls `handle_response_async`

I *think* this will be equivalent to the current setup but avoid having the extra async noise in the operations that don't need async response handling.

Reply:
nice, updated! I left the name for `Operation::handle_response` as-is as I think the signature and callsite make it clear that the method is async